### PR TITLE
[codex] Migrate diarization to pyannote.audio 4

### LINKED ISSUE #7: `--diarize` broken out-of-the-box: 4 compounding dependency issues
Fresh `uv tool install "mlx-qwen3-asr[diarize]"` on macOS produces a working
transcription path but `--diarize` fails with a misleading "gated repo" error
that masks real causes. Debugging revealed 4 stacked issues:

1. **pyannote.audio 4.x incompatibility**  
   `[diarize]` extra resolves to pyannote.audio 4.x, but `diarization.py:267`
   calls `Pipeline.from_pretrained(..., use_auth_token=...)`. This kwarg was
   renamed to `token=` in pyannote 4.0.
   → Fix: rename kwarg, or pin `pyannote.audio<4` in the `[diarize]` extra.

2. **torchaudio 2.8 removed `AudioMetaData`**  
   pyannote.audio 3.x imports `torchaudio.AudioMetaData` which was removed
   in torchaudio 2.8. Default resolution brings torchaudio 2.8.
   → Fix: pin `torchaudio<2.8` (or use pyannote 4.x with fix #1).

3. **huggingface_hub 0.26+ removed `use_auth_token`**  
   pyannote 3.4 internally passes `use_auth_token=` to `hf_hub_download()`,
   which no longer accepts it as of huggingface_hub 0.26.
   → Fix: pin `huggingface_hub<0.26`.

4. **PyTorch 2.6+ weights_only=True default breaks pyannote checkpoints**  
   `torch.load()` default changed to `weights_only=True` in 2.6; pyannote 3.1
   checkpoints contain `TorchVersion` globals that fail unpickling.
   → Fix: pin `torch<2.6` or patch the loader with `weights_only=False`.

**Suggested consolidated fix** in `pyproject.toml`:
```
[project.optional-dependencies]
diarize = [
"pyannote.audio>=3.1,<4",
"torch>=2.2,<2.6",
"torchaudio>=2.2,<2.6",
"huggingface_hub>=0.23,<0.26",
]
```

**Error masking**: `diarization.py:273` wraps the real exception in a
generic "gated repo" RuntimeError. The original exception should be
chained with `raise ... from exc` preserved in the message (or use
`logger.exception` before re-raising) so users can debug without
monkey-patching.

### PR DESCRIPTION
**Pull request by moona3k**
## Summary

Fixes #7 by making pyannote.audio 4.x the supported optional diarization backend for `[diarize]`. Fresh installs now resolve to runtime code that matches the current pyannote auth, model, and output contracts instead of pyannote 3-era behavior.

Core ASR remains MLX-native and torch-free unless users opt into `[diarize]`.

## Root Cause

Issue #7 showed that `[diarize]` allowed pyannote 4 installs through `pyannote.audio>=3.3.2`, while the runtime still assumed pyannote 3 contracts: `use_auth_token`, direct `Annotation` output, and `speaker-diarization-3.1`. That broke the documented fresh-install path and hid actionable dependency/API failures behind confusing gated-model errors.

## Flow

```text
mlx-qwen3-asr --diarize
        |
        v
load pyannote.audio 4 pipeline
        |
        |  token=PYANNOTE_AUTH_TOKEN / HF_TOKEN
        v
pyannote/speaker-diarization-community-1
        |
        v
DiarizeOutput
        |
        |-- exclusive_speaker_diarization  (preferred)
        |-- speaker_diarization            (fallback)
        `-- direct Annotation/list          (legacy/local fallback)
        |
        v
speaker turns -> ASR word/chunk speaker attribution
```

## What Changed

- Pins the optional diarization extra to `pyannote.audio>=4.0.4,<5`.
- Switches the default model to `pyannote/speaker-diarization-community-1`.
- Uses modern `token=` auth, with a legacy-only `use_auth_token=` fallback.
- Rejects `Pipeline.from_pretrained(...) is None` as an initialization failure with a clear root cause.
- Handles pyannote 4 `DiarizeOutput`, preferring exclusive diarization while falling back safely.
- Updates CLI doctor/preflight messages, README setup/troubleshooting, and decision history.
- Adds regression coverage for auth kwargs, output unwrapping, empty exclusive output, root-cause errors, and failed pipeline loads.

## Validation

- `uv pip compile pyproject.toml --extra diarize --python-platform aarch64-apple-darwin --python-version 3.13` resolves `pyannote-audio==4.0.4`, `torchaudio==2.11.0`, `torchcodec==0.11.1`.
- `.venv/bin/python -m pytest tests/test_diarization.py tests/test_cli.py -q` -> 43 passed.
- `.venv/bin/python -m pytest tests/ -q` -> 567 passed, 2 skipped.
- `.venv/bin/ruff check mlx_qwen3_asr tests/test_diarization.py tests/test_cli.py` -> pass.
- `.venv/bin/python scripts/quality_gate.py --mode fast` -> pass.
- Live local integration: `RUN_DIARIZATION_LIVE_INTEGRATION=1 .venv/bin/python -m pytest -q tests/test_diarization_live_integration.py` -> 1 passed.
- GitHub Actions run `24946715422`: `lint`, `test`, and `diarization-integration` all passed.

### CODE REVIEW

File: mlx_qwen3_asr/diarization.py [hunk 1]
```diff
@@ -264,22 +253,36 @@ def _load_pyannote_pipeline() -> object:
...
+        return {}
+    try:
+        params = signature(from_pretrained).parameters
+    except (TypeError, ValueError):
+        return {"token": token}
+    if "token" in params:
+        return {"token": token}
+    if "use_auth_token" in params:
+        return {"use_auth_token": token}
+    return {}
```
[the remark this answers is not in the document]
Comment (moona3k) [reply mlx_qwen3_asr/diarization.py]: Valid. Addressed in c90f758 by defaulting authenticated loads to the modern `token=` kwarg unless the inspected signature is legacy-only (`use_auth_token` without `token`). I also added `test_load_pyannote_pipeline_defaults_to_token_for_kwargs_signature` to cover wrapper-style `**kwargs` signatures.

Validation:
- `.venv/bin/python -m pytest tests/test_diarization.py tests/test_cli.py -q` -> 42 passed
- `.venv/bin/ruff check mlx_qwen3_asr tests/test_diarization.py tests/test_cli.py` -> pass
- `.venv/bin/python -m pytest tests/ -q` -> 557 passed, 2 skipped
- `.venv/bin/python scripts/quality_gate.py --mode fast` -> pass

**Comment by moona3k** [comment]
PR comments addressed/verified after c90f758:

- Gemini inline auth feedback: valid and fixed. `_pyannote_auth_kwargs` now forwards `token=` by default and keeps the legacy `use_auth_token=` path only for legacy-only signatures. Added regression coverage for `**kwargs` wrappers.
- CodeRabbit comment: verified as non-code feedback. It skipped review because this PR is still draft, so there was no CodeRabbit code feedback to address yet. A review can be triggered after the PR is marked ready or with `@coderabbitai review`.
- Additional fresh-eye finding: fixed the empty `exclusive_speaker_diarization` edge case by falling back to regular `speaker_diarization` before the single-speaker fallback. Added regression coverage.

Validation:
- `.venv/bin/python -m pytest tests/test_diarization.py tests/test_cli.py -q` -> 42 passed
- `.venv/bin/ruff check mlx_qwen3_asr tests/test_diarization.py tests/test_cli.py` -> pass
- `.venv/bin/python -m pytest tests/ -q` -> 557 passed, 2 skipped
- `.venv/bin/python scripts/quality_gate.py --mode fast` -> pass

**Comment by moona3k** [comment]
Follow-up on the CodeRabbit auto-reply: verified that it triggered and completed a review run. CI lint/test are green on c90f758.

CodeRabbit produced:
- One optional nitpick about the explicit `torchcodec` doctor check. Verified and left it intentionally because it provides a fast diagnostic for broken or manually modified `[diarize]` environments.
- One docstring-coverage warning in its pre-merge checks. Verified as repo-wide/non-enforced by this project’s local quality gate, not a PR-specific regression.

No additional code changes needed for the CodeRabbit feedback.

**Comment by moona3k** [comment]
Final fresh-eye finding addressed in c90a491.

What changed:
- `_load_pyannote_pipeline()` now treats `Pipeline.from_pretrained(...) is None` as an initialization failure instead of caching/returning a non-callable pipeline.
- The error uses the same gated-model guidance and includes a concrete root cause: `Pipeline.from_pretrained returned None; pyannote could not download or load the pipeline configuration.`
- Added `test_load_pyannote_pipeline_rejects_none_return` to cover pyannote's sentinel-return failure path.

Validation:
- `.venv/bin/python -m pytest tests/test_diarization.py tests/test_cli.py -q` -> 43 passed
- `.venv/bin/ruff check mlx_qwen3_asr tests/test_diarization.py tests/test_cli.py` -> pass
- `.venv/bin/python -m pytest tests/ -q` -> 567 passed, 2 skipped
- `.venv/bin/python scripts/quality_gate.py --mode fast` -> pass
- `git diff --check` -> pass

**Comment by moona3k** [comment]
Live diarization validation is now complete.

Local validation:
- Granted HF access was confirmed by downloading `pyannote/speaker-diarization-community-1` config with the token.
- Installed `[diarize]` locally and confirmed `pyannote.audio==4.0.4`, `torchaudio==2.11.0`, and `torchcodec==0.11.1`.
- `RUN_DIARIZATION_LIVE_INTEGRATION=1 .venv/bin/python -m pytest -q tests/test_diarization_live_integration.py` -> 1 passed.

CI validation:
- Added `PYANNOTE_AUTH_TOKEN` as a repo Actions secret.
- Added `RUN_DIARIZATION_INTEGRATION=1` as a repo Actions variable.
- Reran PR CI run `24946715422`; `diarization-integration`, `lint`, and `test` all passed.
